### PR TITLE
CTU implementation

### PR DESCRIPTION
**Pull request by gerazo**
The final version crushed into one commit with tests.

### CODE REVIEW

File: libcodechecker/analyze/analyzers/analyzer_types.py [hunk 1]
```diff
@@ -246,6 +246,15 @@ def __build_clangsa_config_handler(args, context):
     config_handler.compiler_sysroot = context.compiler_sysroot
     config_handler.system_includes = context.extra_system_includes
     config_handler.includes = context.extra_includes
+
+    if 'ctu_phases' in args:
+        config_handler.ctu_dir = os.path.join(args.output_path,
+                                              args.ctu_dir)
+        config_handler.ctu_in_memory = args.ctu_in_memory
+        config_handler.log_file = args.logfile
+        config_handler.path_env_extra = context.path_env_extra
```
Comment (Xazax-hun) [inline libcodechecker/analyze/analyzers/analyzer_types.py]: What is this and the next entry in config_handler used for?

File: libcodechecker/libhandlers/analyze.py [hunk 2]
```diff
@@ -177,6 +189,45 @@ def add_arguments_to_parser(parser):
...
+        ctu_mutex_group = ctu_opts.add_mutually_exclusive_group()
+        ctu_mutex_group.add_argument('--ctu-all',
+                                     action='store_const',
+                                     const=[True, True],
+                                     dest='ctu_phases',
+                                     default=argparse.SUPPRESS,
+                                     help="Perform cross translation unit "
+                                          "(CTU) analysis (both collect and "
+                                          "analyze phases) using the default "
+                                          "CTU directory as output. "
```
Comment (whisperity) [inline libcodechecker/libhandlers/analyze.py]: The concept of "CTU directory" is no longer explained here, so we should not rely on it. I'll cook up better help messages.

File: libcodechecker/libhandlers/analyze.py [hunk 3]
```diff
@@ -177,6 +189,45 @@ def add_arguments_to_parser(parser):
...
+                                          "At the end of the analysis, this "
+                                          "directory is removed.")
+        ctu_mutex_group.add_argument('--ctu-collect',
+                                     action='store_const',
+                                     const=[True, False],
+                                     dest='ctu_phases',
+                                     default=argparse.SUPPRESS,
+                                     help="Perform the 1st, collect phase of "
+                                          "CTU. Keep CTU directory for "
```
Comment (whisperity) [inline libcodechecker/libhandlers/analyze.py]: > CTU directory
again

File: libcodechecker/libhandlers/analyze.py [hunk 4]
```diff
@@ -177,6 +189,45 @@ def add_arguments_to_parser(parser):
...
+                                          "further use.")
+        ctu_mutex_group.add_argument('--ctu-analyze',
+                                     action='store_const',
+                                     const=[False, True],
+                                     dest='ctu_phases',
+                                     default=argparse.SUPPRESS,
+                                     help="Perform only the 2nd, analyze "
+                                          "phase of CTU. CTU directory should "
```
Comment (whisperity) [inline libcodechecker/libhandlers/analyze.py]: And again.

File: tests/functional/ctu/__init__.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# coding=utf-8
```
Comment (whisperity) [inline tests/functional/ctu/__init__.py]: Why is this file set to `utf-8` encoding?

**Comment by gerazo** [comment]
This is needed to create the correct calling environment for clang. The 
same is used at other places where clang is called.


2017-08-07 10:43 keltezéssel, Gábor Horváth írta:

**Comment by Xazax-hun** [comment]
Oh, I kind of surprised that this change is CTU dependent. What is the reason behind this?

**Comment by gerazo** [comment]
As I remember (I don't see code now), at some CTU related places, we only have the config_handler. I've introduced it when I wanted to call clang to get the arch triplet the same way as it is called when the real thing happens.

File: libcodechecker/libhandlers/analyze.py (hunk 2 above)
Comment (gerazo) [inline libcodechecker/libhandlers/analyze.py]: Thanks. Yes, I've forgot it there.

File: tests/functional/ctu/__init__.py (hunk 5 above)
Comment (gerazo) [inline tests/functional/ctu/__init__.py]: The automatic script created it. I didn't check.

Comment (whisperity) [inline tests/functional/ctu/__init__.py]: And I forgot to realise every file is like that, then. Sorry.

**Comment by whisperity** [comment]
@gerazo If you can't really have access to this code anymore in means of changing it, I'll update the help in a future pull request, is that okay with you, and the rest of us? :slightly_smiling_face:

**Pull request merged**